### PR TITLE
[SPARK-34133][AVRO] Respect case sensitivity when performing Catalyst-to-Avro field matching

### PR DESCRIPTION
**Pull request by xkrogen**
### What changes were proposed in this pull request?
Make the field name matching between Avro and Catalyst schemas, on both the reader and writer paths, respect the global SQL settings for case sensitivity (i.e. case-insensitive by default). `AvroSerializer` and `AvroDeserializer` share a common utility in `AvroUtils` to search for an Avro field to match a given Catalyst field.

### Why are the changes needed?
Spark SQL is normally case-insensitive (by default), but currently when `AvroSerializer` and `AvroDeserializer` perform matching between Catalyst schemas and Avro schemas, the matching is done in a case-sensitive manner. So for example the following will fail:
```scala
      val avroSchema =
        """
          |{
          |  "type" : "record",
          |  "name" : "test_schema",
          |  "fields" : [
          |    {"name": "foo", "type": "int"},
          |    {"name": "BAR", "type": "int"}
          |  ]
          |}
      """.stripMargin
      val df = Seq((1, 3), (2, 4)).toDF("FOO", "bar")

      df.write.option("avroSchema", avroSchema).format("avro").save(savePath)
```

The same is true on the read path, if we assume `testAvro` has been written using the schema above, the below will fail to match the fields:
```scala
df.read.schema(new StructType().add("FOO", IntegerType).add("bar", IntegerType))
  .format("avro").load(testAvro)
```

### Does this PR introduce _any_ user-facing change?
When reading Avro data, or writing Avro data using the `avroSchema` option, field matching will be performed with case sensitivity respecting the global SQL settings.

### How was this patch tested?
New tests added to `AvroSuite` to validate the case sensitivity logic in an end-to-end manner through the SQL engine.

### CODE REVIEW

**Comment by xkrogen** [comment]
cc @dbtsai @viirya @gengliangwang @cloud-fan

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroDeserializer.scala [hunk 1]
```diff
@@ -330,27 +341,27 @@ private[sql] class AvroDeserializer(
     var i = 0
     while (i < length) {
       val sqlField = sqlType.fields(i)
-      val avroField = avroType.getField(sqlField.name)
```
Comment (cloud-fan) [inline external/avro/src/main/scala/org/apache/spark/sql/avro/AvroDeserializer.scala:333]: is the original code `avroType.getField(sqlField.name)` also a O(n) linear lookup?

Comment (xkrogen) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/AvroDeserializer.scala:333]: No, it is a constant-time map lookup (from `org.apache.avro.Schema.RecordSchema`):
```
    public Field getField(String fieldname) {
      if (fieldMap == null)
        throw new AvroRuntimeException("Schema fields not set yet");
      return fieldMap.get(fieldname);
    }
```
I think this is not so performance-critical given that it is executed only once per query? It is not done on a per-record basis.

**Comment by cloud-fan** [comment]
Can we improve the error message in a separated PR? This PR is a bit hard to review as many diffs are not related to the case sensitivity fix.

**Comment by xkrogen** [comment]
Yes, sure. The error message changes did grow fairly large in the end. I will post a separated PR later today.

**Comment by xkrogen** [comment]
Thanks for the suggestion @cloud-fan, I have updated the PR branch, updated the description of the PR/JIRA, and created a new JIRA for the error messages: SPARK-34182

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Since the master branch upgrades Avro to 1.10.1, I re-trigger the tests.

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala [hunk 2]
```diff
@@ -201,4 +201,25 @@ private[sql] object AvroUtils extends Logging {
...
+
+  /**
+   * Extract a single field from `avroFields` which has the same name has `catalystField`,
+   * performing the matching with proper case sensitivity according to [[SQLConf.CASE_SENSITIVE]].
+   *
+   * @param avroFields The fields within which to search for a match.
+   * @param catalystField The catalyst field for which to search for a match.
+   * @return `Some(match)` if a matching Avro field is found, otherwise `None`.
+   */
+  private[avro] def getAvroFieldForCatalystField(
```
Comment (cloud-fan) [inline external/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala:213]: This has nothing to do with catalyst field, only the name, how about
```
def getAvroFieldByName(avroSchema: Schema, name: String): Option[Schema.Field] = {
  val isEqual = SQLConf.get.resolver
  avroSchema.getFields.asScala.filter ...
}
```

File: external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala [hunk 3]
```diff
@@ -1261,6 +1262,58 @@ abstract class AvroSuite
     }
   }
 
+  test("Reading with user provided schema respects case sensitivity for field matching") {
```
Comment (cloud-fan) [inline external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala:1265]: let's put the jira id in the test name: `test("SPARK-xxxx: ....")`

File: external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala [hunk 4]
```diff
@@ -1261,6 +1262,58 @@ abstract class AvroSuite
...
+    assert(withSchema.sameElements(withOutSchema))
+
+    withSQLConf((SQLConf.CASE_SENSITIVE.key, "true")) {
+      val  out = spark.read.format("avro").schema(wrongCaseSchema).load(testAvro).collect()
+      assert(out.forall(_.isNullAt(0)))
+      assert(out.forall(_.isNullAt(1)))
+    }
+  }
+
+  test("Writing with user provided schema respects case sensitivity for field matching") {
```
Comment (cloud-fan) [inline external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala:1285]: ditto

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala (hunk 2 above)
Comment (gengliangwang) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala:213]: +1

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala [hunk 5]
```diff
@@ -201,4 +201,25 @@ private[sql] object AvroUtils extends Logging {
...
+    avroFields: TraversableOnce[Schema.Field],
+    catalystField: StructField): Option[Schema.Field] = {
+    val isEqual: (String, String) => Boolean =
+      if (SQLConf.get.caseSensitiveAnalysis) { _.equals(_) } else { _.equalsIgnoreCase(_) }
+    avroFields.filter(f => isEqual(f.name(), catalystField.name)).toSeq match {
+      case Seq(avroField) => Some(avroField)
+      case Seq() => None
+      case s => throw new IncompatibleSchemaException(
+        s"Searching for ${catalystField.name} in Avro schema gave ${s.size} matches")
```
Comment (gengliangwang) [inline external/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala:222]: Nit: Let's print the Avro field names in the error message.

Comment (gengliangwang) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala:222]: Also, we need a test case for this

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala (hunk 2 above)
Comment (xkrogen) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala:213]: Completely right, good call. Also thanks for the pointer on `SQLConf.resolver`, that is much nicer!

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala (hunk 5 above)
Comment (xkrogen) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala]: Good idea on printing Avro field names! I added a test case as well.

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala [hunk 6]
```diff
@@ -201,4 +203,32 @@ private[sql] object AvroUtils extends Logging {
...
+   *
+   * @param avroSchema The schema in which to search for the field. Must be of type RECORD.
+   * @param name The name of the field to search for.
+   * @return `Some(match)` if a matching Avro field is found, otherwise `None`.
+   * @throws IncompatibleSchemaException if `avroSchema` is not a RECORD or contains multiple
+   *                                     fields matching `name` (i.e., case-insensitive matching
+   *                                     is used and `avroSchema` has two or more fields that have
+   *                                     the same name with difference case).
+   */
+  private[avro] def getAvroFieldByName(avroSchema: Schema, name: String):
```
Comment (gengliangwang) [inline external/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala:219]: As per https://github.com/databricks/scala-style-guide, let's make it:
```
private[avro] def getAvroFieldByName(
    avroSchema: Schema,
    name: String): Option[Schema.Field] = {
```

**Review by gengliangwang — APPROVED** [review]
LGTM except one comment

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala (hunk 6 above)
Comment (xkrogen) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala:219]: I think the way I have it now is valid under the style guide:

> For method declarations, use 4 space indentation for their parameters and put each in each line when the parameters don't fit in two lines. Return types can be either on the same line as the last parameter, or start a new line with 2 space indent.

Example:
```
def newAPIHadoopFile[K, V, F <: NewInputFormat[K, V]](
    path: String,
    fClass: Class[F],
    kClass: Class[K],
    vClass: Class[V],
    conf: Configuration = hadoopConfiguration)
  : RDD[(K, V)] = {
  // method body
}
```

It does seem that the colon is supposed to go on the second line, though. I need to update it anyway and don't have much of an opinion about which is cleaner so I'll go with your proposal.

**Comment by xkrogen** [comment]
Thanks for the reviews @gengliangwang ! Should be all good to go now.

**Comment by cloud-fan** [comment]
thanks, merging to master/3.1!

**Comment by xkrogen** [comment]
Many thanks @cloud-fan !
And thank you @gengliangwang for the reviews!

**Comment by dongjoon-hyun** [comment]
+1, late LGTM. Thank you all!

**Pull request closed without merging**